disnake/state.py

```diff
@@ -784,24 +784,32 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
             self._messages.append(message)
         # we ensure that the channel is a type that implements last_message_id
         if channel:
-            if channel.__class__ in (TextChannel, Thread, VoiceChannel):
+            if type(channel) in (TextChannel, Thread, VoiceChannel):
                 channel.last_message_id = message.id  # type: ignore
-            if (
-                channel.__class__ is Thread
-                and message.type is not MessageType.thread_starter_message
+            # Essentially, messages *don't* count towards message_count, if:
+            # - they're the thread starter message
+            # - or, they're the initial message of a forum channel thread (which uses MessageType.default)
+            # This mirrors the current client and API behavior.
+            if type(channel) is Thread and not (
```

Comment:
Why not `isinstance(channel, Thread)`?

Reply:
FYI we had a discussion about this before
https://discord.com/channels/808030843078836254/913779868985090089/1017172795039432786

Reply:
Why do we want this type check to be strict? Do we have any `Thread` subclasses to worry about this?
Thing is, `isinstance(...)` seems to be almost 1.5 times faster than `type(...) is ...` so that's why I'm asking.

Reply:
> Thing is, `isinstance(...)` seems to be almost 1.5 times faster than `type(...) is ...` so that's why I'm asking.

I can't seem to replicate that:
```py
In [1]: from disnake import TextChannel, Thread

In [2]: textchannel = TextChannel.__new__(TextChannel)

In [3]: thread = Thread.__new__(Thread)

In [4]: %timeit isinstance(thread, Thread)
56.3 ns ± 2.05 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: %timeit type(thread) is Thread
56.4 ns ± 1.12 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit isinstance(textchannel, Thread)
75.1 ns ± 2.15 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit type(textchannel) is Thread
56.8 ns ± 1.5 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
I was already aware that my measurements were not reliable, most likely the speed depends on the cases, have you also tried to measure the execution time with a real Thread? (I searched on the internet in numerous forums and generally, through some measures with the timeit module, the fastest solution is `isinstance`)

Reply:
@shiftinv what times do you get for comparing `__class__` as well?

Reply:
Let's do it properly then, if we're counting nanoseconds :^)
~~`pyperf --duplicate` my beloved~~


```py
$ python test.py -q
__class__ (no match): Mean +- std dev: 27.1 ns +- 2.3 ns
__class__ (match): Mean +- std dev: 27.6 ns +- 4.6 ns
type (no match): Mean +- std dev: 37.0 ns +- 4.0 ns
type (match): Mean +- std dev: 36.7 ns +- 3.4 ns
isinstance (no match): Mean +- std dev: 56.2 ns +- 4.5 ns
isinstance (match): Mean +- std dev: 35.1 ns +- 3.0 ns
```
stddev is pretty high at around 10% of the mean, but comparing these should still be relatively accurate

```py
import pyperf

runner = pyperf.Runner()

setup = """
from disnake import TextChannel, Thread
c = TextChannel.__new__(TextChannel)
t = Thread.__new__(Thread)
"""

kwargs = {"setup": setup, "duplicate": 512}

runner.timeit("__class__ (no match)", "c.__class__ is Thread", **kwargs)
runner.timeit("__class__ (match)", "t.__class__ is Thread", **kwargs)
runner.timeit("type (no match)", "type(c) is Thread", **kwargs)
runner.timeit("type (match)", "type(t) is Thread", **kwargs)
runner.timeit("isinstance (no match)", "isinstance(c, Thread)", **kwargs)
runner.timeit("isinstance (match)", "isinstance(t, Thread)", **kwargs)
```

tldr `isinstance` if the type doesn't match is the slowest, everything else doesn't matter in practice